lib/Backend/Lower.cpp

```diff
@@ -22183,72 +22184,81 @@ Lowerer::GenerateFastArgumentsLdElemI(IR::Instr* ldElem, IR::LabelInstr *labelFa
+                actualParamOpnd = loadInputParamCountInstr->GetDst()->UseWithNewType(TyInt32,this->m_func);
+            }
 
-        if (isInlinee)
-        {
-            if (!hasIntConstIndex)
+            if (hasIntConstIndex)
             {
-                //Runtime check if to make sure length is within the arguments.length range.
-                GenerateCheckForArgumentsLength(ldElem, labelCreateHeapArgs, valueOpnd, actualParamOpnd, Js::OpCode::BrGe_A);
+                //Constant index
+                valueOpnd = IR::IntConstOpnd::New(value, TyInt32, func);
+            }
+            else
+            {
+                //Load valueOpnd from the index
+                valueOpnd =
+                    m_lowererMD.LoadNonnegativeIndex(
+                        indexOpnd,
+                        (
+    #if INT32VAR
+                            indexOpnd->GetType() == TyUint32
+    #else
+                            // On 32-bit platforms, skip the negative check since for now, the unsigned upper bound check covers it
+                            true
+    #endif
+                            ),
+                        labelCreateHeapArgs,
+                        labelCreateHeapArgs,
+                        ldElem);
             }
-        }
-        else
-        {
-            GenerateCheckForArgumentsLength(ldElem, labelCreateHeapArgs, actualParamOpnd, valueOpnd, Js::OpCode::BrLe_A);
-        }
 
-        IR::Opnd *argIndirOpnd = nullptr;
-        if (isInlinee)
-        {
-            argIndirOpnd = GetArgsIndirOpndForInlinee(ldElem, valueOpnd);
+            if (isInlinee)
+            {
+                if (!hasIntConstIndex)
+                {
+                    //Runtime check if to make sure length is within the arguments.length range.
+                    GenerateCheckForArgumentsLength(ldElem, labelCreateHeapArgs, valueOpnd, actualParamOpnd, Js::OpCode::BrGe_A);
+                }
+            }
+            else
+            {
+                GenerateCheckForArgumentsLength(ldElem, labelCreateHeapArgs, actualParamOpnd, valueOpnd, Js::OpCode::BrLe_A);
+            }
+
+            IR::Opnd *argIndirOpnd = nullptr;
+            if (isInlinee)
+            {
+                argIndirOpnd = GetArgsIndirOpndForInlinee(ldElem, valueOpnd);
+            }
+            else
+            {
+                argIndirOpnd = GetArgsIndirOpndForTopFunction(ldElem, valueOpnd);
+            }
+
+            Lowerer::InsertMove(ldElem->GetDst(), argIndirOpnd, ldElem);
+
+            // JMP $done
+            InsertBranch(Js::OpCode::Br, labelFallThru, ldElem);
+            // $labelCreateHeapArgs:
+            ldElem->InsertBefore(labelCreateHeapArgs);
         }
         else
         {
-            argIndirOpnd = GetArgsIndirOpndForTopFunction(ldElem, valueOpnd);
+            IR::Opnd *undef = LoadLibraryValueOpnd(ldElem, LibraryValue::ValueUndefined);
+            Lowerer::InsertMove(ldElem->GetDst(), undef, ldElem);
```

Comment:
This logic is the ride-along for functions that are not (yet) inlined, when attempting to jit `arguments[-n]` before it would always fail to emit the fastpath here which triggered a rejit. Instead now we insert a fast load of undefined.

(Whilst `arguments[-n]` is not the sort of logic we expect in any nice JS it seems silly to force a rejit for it when the per-spec action is simply to load undefined)

Reply:
How does this work if the correct value of arguments[-1] is other than undefined?